browser/ui/views/location_bar/brave_news_location_view.cc

```diff
@@ -10,210 +10,154 @@
-      SetVisible(false);
-      return;
-    }
 
-    auto* contents = tab_strip_->GetActiveWebContents();
-    bool subscribed = false;
-    bool has_feeds = false;
-    absl::optional<BraveNewsTabHelper::FeedDetails> feed;
+  should_show_.Init(brave_news::prefs::kShouldShowToolbarButton,
+                    profile->GetPrefs(),
+                    base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                        base::Unretained(this)));
+  opted_in_.Init(brave_news::prefs::kBraveTodayOptedIn, profile->GetPrefs(),
+                 base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                     base::Unretained(this)));
+  news_enabled_.Init(brave_news::prefs::kNewTabPageShowToday,
+                     profile->GetPrefs(),
+                     base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                         base::Unretained(this)));
+
+  Update();
+}
+
+BraveNewsLocationView::~BraveNewsLocationView() = default;
+
+views::BubbleDialogDelegate* BraveNewsLocationView::GetBubble() const {
+  return bubble_view_;
+}
 
+void BraveNewsLocationView::UpdateImpl() {
+  auto* contents = GetWebContents();
+  BraveNewsTabHelper* tab_helper =
+      contents ? BraveNewsTabHelper::FromWebContents(contents) : nullptr;
+
+  // When the active tab changes, subscribe to notification when
+  // it has found a feed, and update display with current state.
+  if (contents != last_contents_) {
+    // Unobserve old Tab
+    if (last_contents_) {
+      BraveNewsTabHelper::FromWebContents(last_contents_)->RemoveObserver(this);
+    }
+    last_contents_ = contents;
+    // Observe new Tab
     if (contents) {
-      auto* tab_helper = BraveNewsTabHelper::FromWebContents(contents);
-      subscribed = tab_helper->IsSubscribed();
-      has_feeds = !tab_helper->GetAvailableFeeds().empty();
+      tab_helper->AddObserver(this);
     }
-
-    auto image = gfx::CreateVectorIcon(kBraveNewsSubscribeIcon, 16,
-                                       GetIconColor(subscribed));
-    SetImage(ButtonState::STATE_NORMAL, image);
-    SetVisible(has_feeds);
   }
 
-  // views::LabelButton:
-  std::unique_ptr<views::LabelButtonBorder> CreateDefaultBorder()
-      const override {
-    std::unique_ptr<views::LabelButtonBorder> border =
-        LabelButton::CreateDefaultBorder();
-    border->set_insets(gfx::Insets::VH(3, 0));
-    return border;
+  // Don't show the icon if preferences don't allow
+  if (!tab_helper || !should_show_.GetValue() || !news_enabled_.GetValue() ||
+      !opted_in_.GetValue()) {
+    SetVisible(false);
+    return;
   }
 
-  std::u16string GetTooltipText(const gfx::Point& p) const override {
-    return l10n_util::GetStringUTF16(IDS_BRAVE_NEWS_ACTION_VIEW_TOOLTIP);
-  }
+  // Icon color changes if any feeds are being followed
+  bool subscribed = false;
+  subscribed = tab_helper->IsSubscribed();
+  SetIconColor(GetIconColor(subscribed));
```

Comment:
I think collapsing these into one line would be good here.
```suggestion
  SetIconColor(tab_helper->IsSubscribed());
```

Reply:
Yes and looking at this again I changed it to follow the pattern of `UpdateIconColor`